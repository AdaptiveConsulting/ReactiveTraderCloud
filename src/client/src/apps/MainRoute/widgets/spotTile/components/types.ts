import { CurrencyPair, Direction, ServiceConnectionStatus } from 'rt-types'
import { SpotTileData } from '../model'
import { ValidationMessage, NotionalUpdate } from './notional/NotionalInput'
import {
  RfqRequest,
  RfqCancel,
  RfqRequote,
  RfqExpired,
  RfqReject,
  RfqReset,
} from '../model/rfqRequest'

export interface TradingMode {
  symbol: CurrencyPair['symbol']
  mode: 'esp' | 'rfq'
}

export type RfqState = 'none' | 'canRequest' | 'requested' | 'received' | 'expired'

export interface RfqActions {
  request: (rfqActionObj: RfqRequest) => void
  cancel: (rfqActionObj: RfqCancel) => void
  reject: (rfqActionObj: RfqReject) => void
  requote: (rfqActionObj: RfqRequote) => void
  expired: (rfqActionObj: RfqExpired) => void
  reset: (rfqActionObj: RfqReset) => void
}

export interface TileSwitchChildrenProps {
  notional: string
  userError: boolean
}

export interface Props {
  currencyPair: CurrencyPair
  spotTileData: SpotTileData
  executionStatus: ServiceConnectionStatus
  executeTrade: (direction: Direction, rawSpotRate: number) => void
  notional: string
  updateNotional: (notionalUpdate: NotionalUpdate) => void
  resetNotional: () => void
  inputDisabled: boolean
  inputValidationMessage: ValidationMessage
  tradingDisabled: boolean
  chartData?: []
  rfq: RfqActions
  displayCurrencyChart?: () => void
}
