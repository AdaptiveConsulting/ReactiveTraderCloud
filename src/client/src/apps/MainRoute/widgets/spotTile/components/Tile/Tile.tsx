import React from 'react'
import { CurrencyPair, Direction, ServiceConnectionStatus } from 'rt-types'
import {
  ExecuteTradeRequest,
  SpotTileData,
  createTradeRequest,
  TradeRequest,
} from '../../model/index'
import SpotTile from '../SpotTile'
import { AnalyticsTile } from '../analyticsTile/index'
import { TileViews } from '../../../workspace/workspaceHeader/index'
import { TileSwitchChildrenProps, TradingMode, RfqActions } from '../types'
import { ValidationMessage } from '../notional/NotionalInput'
import {
  getDefaultInitialNotionalValue,
  getDerivedStateFromProps,
  getNumericNotional,
  getDerivedStateFromUserInput,
  isValueInRfqRange,
  getDefaultNotionalValue,
} from './TileBusinessLogic'
import { getConstsFromRfqState } from '../../model/spotTileUtils'
import { CurrencyPairNotional } from '../../model/spotTileData'
import { NotionalUpdate } from '../../model/spotTileData'

export interface TileProps {
  currencyPair: CurrencyPair
  spotTileData: SpotTileData
  executionStatus: ServiceConnectionStatus
  executeTrade: (tradeRequestObj: ExecuteTradeRequest) => void
  setTradingMode: (tradingMode: TradingMode) => void
  displayCurrencyChart?: () => void
  tileView?: TileViews
  children: ({ notional, userError }: TileSwitchChildrenProps) => JSX.Element
  rfq: RfqActions
  updateNotional: (notionalUpdate: CurrencyPairNotional) => void
}

export interface TileState {
  notional: string
  inputDisabled: boolean
  inputValidationMessage: ValidationMessage
  tradingDisabled: boolean
  canExecute: boolean
}

class Tile extends React.PureComponent<TileProps, TileState> {
  state: TileState = {
    notional: getDefaultInitialNotionalValue(this.props.currencyPair),
    inputValidationMessage: null,
    inputDisabled: false,
    tradingDisabled: false,
    canExecute: true,
  }

  tileComponents = {
    [TileViews.Normal]: SpotTile,
    [TileViews.Analytics]: AnalyticsTile,
  }

  // State management derived from props
  static getDerivedStateFromProps(nextProps: TileProps, prevState: TileState) {
    const { setTradingMode, spotTileData, currencyPair } = nextProps
    return {
      ...getDerivedStateFromProps(nextProps, prevState),
      ...getDerivedStateFromUserInput({
        prevState,
        notionalUpdate: spotTileData.notional,
        spotTileData,
        actions: { setTradingMode },
        currencyPair,
      }),
    }
  }

  // We handle the case where the initial Notional value would
  // be in the RFQ range.
  componentDidMount() {
    const {
      spotTileData: { rfqState },
      setTradingMode,
      currencyPair: { symbol },
    } = this.props
    const { notional } = this.state
    const { isRfqStateNone } = getConstsFromRfqState(rfqState)

    if (isRfqStateNone && isValueInRfqRange(notional)) {
      setTradingMode({ symbol, mode: 'rfq' })
    }
  }

  executeTrade = (direction: Direction, rawSpotRate: number) => {
    const {
      currencyPair,
      executeTrade,
      rfq,
      spotTileData: { rfqState },
    } = this.props
    const { notional } = this.state
    const { isRfqStateReceived } = getConstsFromRfqState(rfqState)
    const tradeRequestObj: TradeRequest = {
      direction,
      currencyBase: currencyPair.base,
      symbol: currencyPair.symbol,
      notional: getNumericNotional(notional),
      rawSpotRate,
    }
    executeTrade(createTradeRequest(tradeRequestObj))

    // After executing the trade we need to
    // reset our RFQ state if necessary.
    if (isRfqStateReceived) {
      rfq.reset({ currencyPair })
    }
  }

  updateNotional = (notionalUpdate: NotionalUpdate) => {
    const { currencyPair, updateNotional } = this.props

    updateNotional({
      currencyPair: currencyPair.symbol,
      notional: notionalUpdate,
    })
  }

  resetNotional = () => {
    const { setTradingMode, currencyPair, updateNotional } = this.props
    const defaultNotional = getDefaultNotionalValue()
    const isInRfqRange = isValueInRfqRange(defaultNotional)
    setTradingMode({ symbol: currencyPair.symbol, mode: isInRfqRange ? 'rfq' : 'esp' })
    updateNotional({
      currencyPair: currencyPair.symbol,
      notional: {
        value: defaultNotional,
      },
    })
  }

  render() {
    const {
      children,
      currencyPair,
      spotTileData,
      executionStatus,
      tileView,
      rfq,
      displayCurrencyChart,
    } = this.props
    const {
      inputDisabled,
      inputValidationMessage,
      tradingDisabled,
      canExecute,
      notional,
    } = this.state
    const { rfqState } = spotTileData
    const { isRfqStateCanRequest, isRfqStateNone } = getConstsFromRfqState(rfqState)

    const TileViewComponent =
      tileView && (isRfqStateNone || isRfqStateCanRequest)
        ? this.tileComponents[tileView]
        : SpotTile

    return (
      <TileViewComponent
        currencyPair={currencyPair}
        spotTileData={spotTileData}
        executeTrade={this.executeTrade}
        executionStatus={executionStatus}
        notional={notional}
        updateNotional={this.updateNotional}
        resetNotional={this.resetNotional}
        inputDisabled={inputDisabled}
        inputValidationMessage={inputValidationMessage}
        tradingDisabled={tradingDisabled || !canExecute}
        rfq={rfq}
        displayCurrencyChart={displayCurrencyChart}
      >
        {children({
          notional,
          userError: Boolean(inputValidationMessage),
        })}
      </TileViewComponent>
    )
  }
}

export default Tile
