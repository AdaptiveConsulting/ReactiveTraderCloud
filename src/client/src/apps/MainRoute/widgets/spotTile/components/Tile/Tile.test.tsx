import React from 'react'
import { renderWithTheme } from '../../../../../../../__tests__/helpers'
import Tile, { TileProps, TileState } from './Tile'
import { currencyPair } from '../test-resources/spotTileProps'
import { ServiceConnectionStatus } from 'rt-types'
import { TileViews } from '../../../workspace/workspaceHeader'
import { DerivedStateFromUserInput } from './TileBusinessLogic'
import { TradingMode } from '../types'
import { PriceMovementTypes } from '../../model/priceMovementTypes'

const prevState: TileState = {
  canExecute: true,
  inputDisabled: false,
  inputValidationMessage: null,
  notional: '1,000,000',
  tradingDisabled: false,
}

const defaultParams: DerivedStateFromUserInput = {
  actions: {
    setTradingMode: (tradingMode: TradingMode) => {},
  },
  prevState,
  notionalUpdate: {
    type: 'blur',
    value: '1,000,000',
  },
  spotTileData: {
    currencyChartIsOpening: false,
    historicPrices: [],
    isTradeExecutionInFlight: false,
    lastTradeExecutionStatus: null,
    price: {
      ask: 1.48364,
      bid: 1.4835,
      creationTimestamp: 694779224112175,
      mid: 1.48357,
      priceMovementType: 'Down' as PriceMovementTypes,
      symbol: 'EURCAD',
      valueDate: '2019-04-07T00:00:00Z',
    },
    rfqPrice: null,
    rfqState: 'none',
    rfqTimeout: null,
  },
  currencyPair,
}

const defaultTileProps: TileProps = {
  children: jest.fn(),
  currencyPair,
  executeTrade: () => {},
  setTradingMode: () => {},
  executionStatus: 'CONNECTED' as ServiceConnectionStatus,
  rfq: {
    cancel: () => {},
    expired: () => {},
    reject: () => {},
    request: () => {},
    requote: () => {},
    reset: () => {},
  },
  spotTileData: defaultParams.spotTileData,
  tileView: 'Normal' as TileViews,
}

test('Snapshot, state derived from props, defaults, RFQ none, should be able to excute', () => {
  const component = renderWithTheme(<Tile {...defaultTileProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, DISCONNECTED', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    executionStatus: 'DISCONNECTED' as ServiceConnectionStatus,
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, in trade', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    spotTileData: {
      ...defaultTileProps.spotTileData,
      isTradeExecutionInFlight: true,
    },
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, RFQ requested', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    spotTileData: {
      ...defaultTileProps.spotTileData,
      rfqState: 'requested',
    },
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, RFQ received', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    spotTileData: {
      ...defaultTileProps.spotTileData,
      rfqState: 'received',
      rfqPrice: {
        ask: 1.48364,
        bid: 1.4835,
        creationTimestamp: 694779224112175,
        mid: 1.48357,
        priceMovementType: 'Down' as PriceMovementTypes,
        symbol: 'EURCAD',
        valueDate: '2019-04-07T00:00:00Z',
      },
    },
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, RFQ canRequest', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    spotTileData: {
      ...defaultTileProps.spotTileData,
      rfqState: 'canRequest',
    },
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Snapshot, state derived from props, RFQ expired', () => {
  const nextProps: TileProps = {
    ...defaultTileProps,
    spotTileData: {
      ...defaultTileProps.spotTileData,
      rfqState: 'expired',
      rfqPrice: {
        ask: 1.48364,
        bid: 1.4835,
        creationTimestamp: 694779224112175,
        mid: 1.48357,
        priceMovementType: 'Down' as PriceMovementTypes,
        symbol: 'EURCAD',
        valueDate: '2019-04-07T00:00:00Z',
      },
    },
  }
  const component = renderWithTheme(<Tile {...nextProps} />)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
