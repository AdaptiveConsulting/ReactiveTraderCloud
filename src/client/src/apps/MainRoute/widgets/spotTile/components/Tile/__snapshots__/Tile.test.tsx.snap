// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot, state derived from props, DISCONNECTED 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <button
          className="sc-EHOje gCbQHr"
          direction="Sell"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                SELL
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <button
          className="sc-EHOje gCbQHr"
          direction="Buy"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                BUY
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb ixCVbX"
            disabled={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, RFQ canRequest 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <div
          className="sc-jlyJG djTXEU"
        >
          <i
            className="fas fa-ban fa-flip-horizontal sc-gipzik hRaVNo"
          />
          Streaming price unavailable
        </div>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <div
          className="sc-jlyJG djTXEU"
        >
          <i
            className="fas fa-ban fa-flip-horizontal sc-gipzik hRaVNo"
          />
          Streaming price unavailable
        </div>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb fauXdE"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
          <button
            className="sc-ifAKCX gROrJN"
            onClick={[Function]}
          >
            <i
              className="fas fa-redo fa-flip-horizontal"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, RFQ expired 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <button
          className="sc-EHOje gCbQHr"
          direction="Sell"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                SELL
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
          <div
            className="sc-VigVT cLQaHF"
          >
            Expired
          </div>
        </button>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <button
          className="sc-EHOje gCbQHr"
          direction="Buy"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                BUY
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
          <div
            className="sc-VigVT cLQaHF"
          >
            Expired
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb fauXdE"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
          <button
            className="sc-ifAKCX gROrJN"
            onClick={[Function]}
          >
            <i
              className="fas fa-redo fa-flip-horizontal"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, RFQ received 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <button
          className="sc-EHOje ixrtnE"
          direction="Sell"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi crpdKd"
            disabled={false}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                SELL
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <button
          className="sc-EHOje iHiLlf"
          direction="Buy"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi crpdKd"
            disabled={false}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                BUY
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb ixCVbX"
            disabled={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, RFQ requested 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <div
          className="sc-jlyJG djTXEU"
        >
          <div
            className="sc-csuQGl kfyPEl"
          >
            <svg
              height={14}
              width={14}
            >
              <rect
                className="sc-kkGfuU isRsVR"
                height={10.5}
                order={0}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={0.25}
              />
              <rect
                className="sc-kkGfuU kjqbOX"
                height={10.5}
                order={1}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={3.875}
              />
              <rect
                className="sc-kkGfuU hcoFxx"
                height={10.5}
                order={2}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={7.5}
              />
              <rect
                className="sc-kkGfuU ffvTAs"
                height={10.5}
                order={3}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={11.125}
              />
            </svg>
          </div>
          Awaiting price
        </div>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <div
          className="sc-jlyJG djTXEU"
        >
          <div
            className="sc-csuQGl kfyPEl"
          >
            <svg
              height={14}
              width={14}
            >
              <rect
                className="sc-kkGfuU isRsVR"
                height={10.5}
                order={0}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={0.25}
              />
              <rect
                className="sc-kkGfuU kjqbOX"
                height={10.5}
                order={1}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={3.875}
              />
              <rect
                className="sc-kkGfuU hcoFxx"
                height={10.5}
                order={2}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={7.5}
              />
              <rect
                className="sc-kkGfuU ffvTAs"
                height={10.5}
                order={3}
                speed={0.8}
                type="secondary"
                width={2.625}
                x={11.125}
              />
            </svg>
          </div>
          Awaiting price
        </div>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb ixCVbX"
            disabled={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, defaults, RFQ none, should be able to excute 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <button
          className="sc-EHOje kkdSuS"
          direction="Sell"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi crpdKd"
            disabled={false}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                SELL
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <button
          className="sc-EHOje ciDDPg"
          direction="Buy"
          disabled={false}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi crpdKd"
            disabled={false}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                BUY
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb fauXdE"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Snapshot, state derived from props, in trade 1`] = `
<div
  className="sc-eqIVtm sc-fAjcbJ iquZYL"
>
  <div
    className="sc-jhAzac spot-tile sc-gisBJw cgJjhx"
  >
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-frDJqD fAbCDq"
      >
        <div
          className="sc-dVhcbM esJDHn"
        >
          <div
            className="sc-hzDkRC hefQwu"
          >
            EUR/USD
          </div>
          <div
            className="sc-bRBYWo jmSJAL"
          >
            SPT (07 APR)
          </div>
        </div>
      </div>
      <div
        className="sc-iRbamj iTHPiO"
      >
        <button
          className="sc-EHOje gCbQHr"
          direction="Sell"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                SELL
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
        <div
          className="sc-jzJRlG iKkJoh"
        >
          <i
            aria-hidden="true"
            className="fas fa-caret-up sc-jTzLTM gkTtVL"
            color="green"
          />
          <div
            className="sc-fjdhpX flAxeK"
          >
            0.0
          </div>
          <i
            aria-hidden="true"
            className="fas fa-caret-down sc-jTzLTM TMXdn"
            color="red"
          />
        </div>
        <button
          className="sc-EHOje gCbQHr"
          direction="Buy"
          disabled={true}
          onClick={[Function]}
        >
          <div
            className="sc-gZMcBi egNMcT"
            disabled={true}
          >
            <div
              className="sc-gqjmRU cKkifn"
            >
              <div
                className="sc-bZQynM sc-gzVnrw lDSPw"
              >
                BUY
              </div>
              <div
                className="sc-bZQynM sc-htoDjs duRaET"
              >
                1.
              </div>
            </div>
            <div
              className="sc-bZQynM sc-dnqmqq fBMQfa"
            >
              48
            </div>
            <div
              className="sc-bZQynM sc-iwsKbI CPNEg"
            >
              4
            </div>
          </div>
        </button>
      </div>
    </div>
    <div
      className="sc-caSCKo jjNSwx"
    >
      <div
        className="sc-kjoXOD bvCYDb"
      >
        <div
          className="sc-htpNat idZgat"
        >
          <span
            className="sc-bdVaJa bxrsSL"
          >
            EUR
          </span>
          <input
            className="sc-bxivhb ixCVbX"
            disabled={true}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="text"
            value="1,000,000"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
